internal/buildpack/ruby.go

```diff
@@ -3,231 +3,130 @@ package buildpack
 import (
 	"context"
 	"fmt"
-	"net/http"
-	"os"
-	"os/exec"
-	"path/filepath"
 	"strings"
 
-	"github.com/johnewart/archiver"
-	"github.com/matishsiao/goInfo"
-
+	"github.com/yourbase/yb/internal/biome"
 	"github.com/yourbase/yb/internal/ybdata"
 	"github.com/yourbase/yb/plumbing"
-	"gopkg.in/src-d/go-git.v4"
+	"github.com/yourbase/yb/types"
 	"zombiezen.com/go/log"
 )
 
-const rubyDownloadTemplate = "https://yourbase-build-tools.s3-us-west-2.amazonaws.com/ruby/ruby-{{ .Version }}-{{ .OS }}-{{ .Arch }}-{{ .OsVersion }}.{{ .Extension }}"
-
-func (bt rubyBuildTool) downloadURL(ctx context.Context) string {
-	extension := "tar.bz2"
-	osVersion := OSVersion()
-	arch := Arch()
-
-	if arch == "amd64" {
-		arch = "x86_64"
-	}
-
-	operatingSystem := OS()
-	if operatingSystem == "darwin" {
-		operatingSystem = "Darwin"
+// TODO: Install libssl-dev (or equivalent / warn) and zlib-dev based on platform
+
+func installRuby(ctx context.Context, sys Sys, spec types.BuildpackSpec) (biome.Environment, error) {
+	rbenvDir := sys.Biome.JoinPath(sys.Biome.Dirs().Tools, "rbenv")
+	rubyDir := sys.Biome.JoinPath(rbenvDir, "versions", spec.Version())
+	gemsDir := sys.Biome.JoinPath(sys.Biome.Dirs().Home, "rubygems")
+	env := biome.Environment{
+		Vars: map[string]string{
+			"GEM_HOME": gemsDir,
+		},
+		PrependPath: []string{
+			sys.Biome.JoinPath(gemsDir, "bin"),
+			sys.Biome.JoinPath(rubyDir, "bin"),
+		},
 	}
 
-	if operatingSystem == "linux" {
-		operatingSystem = "Linux"
-	}
-
-	if operatingSystem == "windows" {
-		extension = "zip"
+	// If directory already exists, then use it.
+	if _, err := biome.EvalSymlinks(ctx, sys.Biome, rubyDir); err == nil {
+		log.Infof(ctx, "Ruby v%s located in %s", spec.Version(), rubyDir)
+		return env, nil
 	}
 
+	// Try to download a YourBase pre-built binary.
+	desc := sys.Biome.Describe()
+	const template = "https://yourbase-build-tools.s3-us-west-2.amazonaws.com/ruby/ruby-{{ .Version }}-{{ .OS }}-{{ .Arch }}-{{ .DistroCodename }}.tar.bz2"
 	data := struct {
-		OS        string
-		OsVersion string
-		Arch      string
-		Version   string
-		Extension string
+		Version        string
+		OS             string
+		DistroCodename string
```

Comment:
It gets set a little farther down, because we only care about this data if it's a supported OS/arch combo. FWIW, the unit test uses the precompiled download.